spacy_experimental/biaffine_parser/arc_predicter.pyx

```diff
@@ -300,3 +312,28 @@ def sents2lens(docs: List[Doc], *, ops: Ops) -> Ints1d:
             lens.append(sent.end - sent.start)
 
     return ops.asarray1i(lens)
+
+def split_lazily(docs: List[Doc], *, ops: Ops, max_length: int, senter: SentenceRecognizer, is_train: bool) -> Ints1d:
+    lens = []
+    for doc in docs:
+        activations = doc.activations.get(senter.name, None)
+        if activations is None:
+            raise ValueError("Lazy splitting requires `senter` with `save_activations` enabled.\n"
+                             "During training, `senter` must be in the list of annotating components")
+        scores = activations['probabilities']
+        split_recursive(scores[:,1], ops, max_length, lens)
+
+    assert sum(lens) == sum([len(doc) for doc in docs])
+
+    return ops.asarray1i(lens)
+
+def split_recursive(scores, ops, max_length, lengths):
```

Comment:
Is this going to be re-written to non-recursive as we discussed due to not being able to deal with the oversplitting during the beginning of training or is there another fix?

Reply:
Or its coming?

Reply:
I'll do that in a separate PR. I think it's nice to have the clean, recursive version somewhere in the history.